ludwig/utils/fs_utils.py

```diff
@@ -126,11 +130,13 @@ def abspath(url):
     return os.path.abspath(url)
 
 
+@default_retry()
```

Comment:
I'm a little bit concerned about adding blanket retries for all these functions. If we set the retry too low, it could have little affect if there's a network partition lasting a minute or more. If it's too high, it could take a while for a real failure to bubble up to the users. Do we have some indication of what types of errors are surfaced when a connection issue arises?

There is one open issue in fsspec [here](https://github.com/fsspec/filesystem_spec/issues/776) that indicates there might be something for S3 and GCS we can use to implement retry logic at the filesystem level. But not sure what those params are or how are they are configured.

Reply:
I looked through the `s3fs` and `gcfs` libraries.

For `s3fs`:
It seems like there's a default retry for many (all?) `s3fs` methods. The retry policy is exponential backoff, with 5 requests with up to ~0.8 seconds of backoff between requests. You can check it out [here](https://s3fs.readthedocs.io/en/latest/_modules/s3fs/core.html#) and cmd+F `async def _error_wrapper`. Within the same file, there's also `S3_RETRYABLE_ERRORS ` which we can use if needed.

For `gcfs`:
Similar behavior, except that the max backoff is 32 seconds. Also seems like it applies to most if not all methods of the class. You can take a look at how retry behavior is configured [here](https://github.com/fsspec/gcsfs/blob/b4eabc01fe5d48ff5511d4fbf543fcbd55435acd/gcsfs/retry.py#L110). 

Given this info:
1. I'm leaning towards just removing retry decorators from any `fsspec` method. However,
2. It's unclear if all filesystems implement their own retry methods. And,
3. The retry policies seem to vary across filesystem, which may be ok I guess?

Wdyt?

Reply:
Can we configure the retry policy for s3 and gcs via environment variables? If we could do the same for mlfow, that would be ideal, I think.

Reply:
@tgaddair Done!